platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ExportMigrationContextImpl.java

```diff
@@ -0,0 +1,278 @@
+import com.google.common.collect.ImmutableMap;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * The export migration context keeps track of exported migration entries for a given migratable
+ * while processing an export migration operation.
+ */
+public class ExportMigrationContextImpl extends MigrationContextImpl
+        implements ExportMigrationContext, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportMigrationContextImpl.class);
+
+    /**
+     * Holds exported migration entries keyed by the exported path.
+     */
+    private final Map<Path, ExportMigrationEntryImpl> entries = new TreeMap<>();
+
+    /**
+     * Holds migration entries referenced from system properties keyed by the property name.
+     */
+    private final Map<String, ExportMigrationSystemPropertyReferencedEntryImpl> systemProperties =
+            new TreeMap<>();
+
+    private final ExportMigrationReportImpl report;
+
+    private final ZipOutputStream zos;
+
+    private volatile OutputStream currentOutputStream;
+
+    /**
+     * Creates a new migration context for an export operation.
+     *
+     * @param report     the migration report where to record warnings and errors
+     * @param migratable the migratable this context is for
+     * @param zos        the output stream for the zip file being generated
+     * @throws IllegalArgumentException if <code>report</code>, <code>migratable</code>, <code>zos</code>
+     *                                  is <code>null</code>
+     * @throws java.io.IOError          if unable to determine ${ddf.home}
+     */
+    ExportMigrationContextImpl(MigrationReport report, Migratable migratable, ZipOutputStream zos) {
+        super(report,
+                migratable,
+                ExportMigrationContextImpl.validateNotNull(migratable, "invalid null migratable")
+                        .getVersion());
+        Validate.notNull(zos, "invalid null zip output stream");
+        this.report = new ExportMigrationReportImpl(report, migratable);
+        this.zos = zos;
+    }
+
+    private static <T> T validateNotNull(T t, String msg) {
+        Validate.notNull(t, msg);
+        return t;
+    }
+
+    @Override
+    public ExportMigrationReportImpl getReport() {
+        return report;
+    }
+
+    @Override
+    public Optional<ExportMigrationEntry> getSystemPropertyReferencedEntry(String name,
+            BiPredicate<MigrationReport, String> validator) {
+        Validate.notNull(name, "invalid null system property name");
+        Validate.notNull(validator, "invalid null validator");
+        final ExportMigrationSystemPropertyReferencedEntryImpl me = systemProperties.get(name);
+
+        if (me != null) {
+            return Optional.of(me);
+        }
+        final String val = System.getProperty(name);
+
+        if (!validator.test(report, val)) {
+            return Optional.empty();
+        } else if (val == null) {
```

Comment:
❓ Since the order in which things are checked is part of the contract, i.e., we can't just check for `null` first and skip the call to the `validator`, should we add a comment and refer to the API documentation so people are not tempted to "optimize" this code?

I actually almost made a comment on it before checking the JavaDoc.